Lib/idlelib/editor.py

```diff
@@ -1465,6 +1437,37 @@ def inner(offset, _startindex=startindex,
             return _icis(_startindex + "+%dc" % offset)
         return inner
 
+    def parser(self, index="insert", lineend=""):
+        "Create a PyParse instance and find the last statement."
+        text = self.text
+        p = pyparse.Parser(self.indentwidth, self.tabwidth)
+        lineno = self.getlineno(index)
+        stopatindex = index if not lineend else f"{lineno}.end"
+
+        if self.prompt_last_line:
+            r = text.tag_prevrange("console", index)
+            startatindex = r[1] if r else "1.0"
```

Comment:
In the shell, it looks like it will always find the last `console` tag, which is marked from the beginning to the end of the prompt.  For example, in a new shell window, typing `>>> def hello(` without pressing Enter shows `r` to be `(3.0, 3.4)`. 

I wasn't able to think of a scenario where the prompt would be missing, so I'm not sure if the default of `1.0` would ever be used.

Reply:
With the shell moved to a sidebar (PR GH-15474), there's still a "console" tag on the newline (`\n`) before the beginning of the current input block, making code such as this continue to "just work".

Reply:
While we're refactoring this, I think it's a great opportunity to remove the shell-specific code from the EditorWindow class and into PyShell.